subprojects/language-native/src/main/java/org/gradle/language/plugins/NativeBasePlugin.java

```diff
@@ -164,6 +161,15 @@ private void addLifecycleTasks(final TaskContainer tasks, final SoftwareComponen
                         tasks.named(LifecycleBasePlugin.ASSEMBLE_TASK_NAME, task -> task.dependsOn(outputs));
                     }
                 });
+
+                tasks.named(LifecycleBasePlugin.ASSEMBLE_TASK_NAME, task -> {
+                    task.dependsOn((Callable) () -> {
+                        if (component.getBinaries().get().isEmpty()) {
```

Comment:
This is a bit too coincidental. We might add another reason why there might be no binaries, but for which it is ok to run `assemble`. I would instead actually check that the component does not target the current machine, by looking at its target machines.